src/app/admin/edit-product/edit-product.component.ts

```diff
@@ -0,0 +1,107 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  OnDestroy,
+  OnInit,
+} from '@angular/core';
+import {
+  FormBuilder,
+  FormControl,
+  FormGroup,
+  Validators,
+} from '@angular/forms';
+import { ActivatedRoute, Router } from '@angular/router';
+
+import { BehaviorSubject, Subject } from 'rxjs';
+import { finalize, takeUntil } from 'rxjs/operators';
+
+import { Product } from '../../products/product.interface';
+import { ProductsService } from '../../products/products.service';
+
+@Component({
+  selector: 'app-edit-product',
+  templateUrl: './edit-product.component.html',
+  styleUrls: ['./edit-product.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class EditProductComponent implements OnInit, OnDestroy {
+  form: FormGroup;
+  product: Product | null = null;
+  requestInProgress = false;
+
+  loaded$ = new BehaviorSubject(false);
+
+  countCtrl: FormControl;
+  descriptionCtrl: FormControl;
+  priceCtrl: FormControl;
+  titleCtrl: FormControl;
+
+  private readonly onDestroy$: Subject<void> = new Subject();
+
+  constructor(
+    private readonly activatedRoute: ActivatedRoute,
+    private readonly fb: FormBuilder,
+    private readonly productsService: ProductsService,
+    private readonly router: Router
+  ) {
+    this.titleCtrl = this.fb.control(null, Validators.required);
+    this.descriptionCtrl = this.fb.control(null, Validators.required);
+    this.priceCtrl = this.fb.control(null, Validators.required);
+    this.countCtrl = this.fb.control(null, Validators.required);
+    this.form = this.fb.group({
+      title: this.titleCtrl,
+      description: this.descriptionCtrl,
+      price: this.priceCtrl,
+      count: this.countCtrl,
+    });
```

Comment:
this.fb.group({
title: ['', Validators.required]
...})

Reply:
The separate form controls are used in a template is order to simplify it and compute the `touched` state on a per-control bases. Do you want to get rid of them?

Reply:
Better use getters for retrieving controls or form.get() in template

Reply:
Done